### ISSUE TITLE
Infinite loop (browser crashes) in Next 13

### ISSUE THREAD

**Issue by victorbianconi** [TanStack/query#5532]
### Describe the bug

Can't get React Query to work in Next 13 (even on client components). The app crashes totally and if I log the query it's basically looped indefinitely. 

### Your minimal, reproducible example

Code

### Steps to reproduce

"use client";

import { useQuery } from "@tanstack/react-query";

const fetchGens = async (pageParam) => {
  const response = await fetch("/api/generations/my?pageParam=" + pageParam);
  const data = await response.json(); // Extract the JSON data from the response
  console.log(data);
  return data;
};

export default async function Generations({ gens }) {
  const {
    fetchNextPage,
    fetchPreviousPage,
    hasNextPage,
    hasPreviousPage,
    isFetchingNextPage,
    isFetchingPreviousPage,
    data,
  } = useQuery({
    queryKey: ["generations"],
    queryFn: ({ pageParam = 0 }) => fetchGens(pageParam),
    getNextPageParam: (lastPage, allPages) => lastPage.nextCursor,
    getPreviousPageParam: (firstPage, allPages) => firstPage.prevCursor,
    onSuccess: (data) => {
      console.log(data)
    },
  });

  return (
    <main className="flex flex-col min-h-[100vh] items-center py-[4rem]">
      {" "}
      <div className="w-[80vw] mobile:w-[90vw] flex flex-col">
        {" "}
        <h1 className="text-3xl text-white font-bold mb-[1.6rem]">
          Generations
        </h1>
        <div className="grid grid-cols-3"> </div>
      </div>
    </main>
  );
}

### Expected behavior

It should not loop nor crash. or

### How often does this bug happen?

Every time

### Platform

Windows, Chrome

### Tanstack Query adapter

react-query

### TanStack Query version

v4.29.12

**Comment by TkDodo** [comment]
please show a runnable reproduction of your issue, or we'll close this issue.

**Comment by victorbianconi** [comment]
I don't have an example right now but this is the code. Not sure how to format properly here. If I use the same code in a normal react app, it works. So I assume it must be related to Next.

```
"use client";

import { useInfiniteQuery } from "@tanstack/react-query";

const fetchGens = async (pageParam) => {
  const response = await fetch("/api/generations/my?pageParam=" + pageParam);
  const data = await response.json(); // Extract the JSON data from the response
  console.log(data);
  return data;
};

export default async function Generations({ gens }) {
  const {
    fetchNextPage,
    fetchPreviousPage,
    hasNextPage,
    hasPreviousPage,
    isFetchingNextPage,
    isFetchingPreviousPage,
    data,
  } = useInfiniteQuery({
    queryKey: ["generations"],
    queryFn: ({ pageParam = 0 }) => fetchGens(pageParam),
    getNextPageParam: (lastPage, allPages) => lastPage.nextCursor,
    getPreviousPageParam: (firstPage, allPages) => firstPage.prevCursor,
    onSuccess: (data) => {
      console.log(data)
    },
  });

  return (
    <main className="flex flex-col min-h-[100vh] items-center py-[4rem]">
      {" "}
      <div className="w-[80vw] mobile:w-[90vw] flex flex-col">
        {" "}
        <h1 className="text-3xl text-white font-bold mb-[1.6rem]">
          Generations
        </h1>
        <div className="grid grid-cols-3"> </div>
      </div>
    </main>
  );
}
```

**Comment by victorbianconi** [comment]
Here you go: https://codesandbox.io/p/sandbox/kind-snow-04rdx0
I hope this suffices. The API obviously doesn't work but I noticed it still loops and crashes. 
Let me know if this is enough info for you.

**Comment by TkDodo** [comment]
Idk but that's exactly the reason why I always demand a reproduction. It's 95% not the libs fault, but something in your surroundings.

**Comment by victorbianconi** [comment]
> Idk but that's exactly the reason why I always demand a reproduction. It's 95% not the libs fault, but something in your surroundings.

I migrated back to Next 12.3.2 and it works just fine. Must have been the latest version from Next.

**Comment by haoliangwu-sbi-dalian** [comment]
I also met a similar issue related to the next13, I will try to create a reproduction tomorrow(and is it better to create another issue to track it maybe?). 

I can provide a .gif to show the issue right now only, see:
![chrome-capture-2023-5-13](https://github.com/TanStack/query/assets/131343689/a06ae229-a3f5-4576-abfc-c5606d911a7f)

Basically, I figured out the root cause is maybe related to suspense mode, as long as the `useQuery`'s fetcher failed **the first time** when the component was mounted, the `useQuery` was trapped in an infinite loop. If the fetcher just failed the second time or further but succeeded at the first time, it wasn't trapped.

The expected behavior should be the error was captured by ErrorBoundray, but it seems like the Suspense always try to re-render the component which contains the failed useQuery hook.

**Comment by Justinfan827** [comment]
I don't think you can have client components using the 'async' keyword: https://github.com/facebook/react/issues/26801

**Comment by TkDodo** [comment]
good catch!
